File: mobile/lib/models/notification.dart
/// One message in the user's inbox -- mirrors the backend's
/// NotificationOut exactly.
///
/// `source` is provenance, not behaviour: "manual" for a message an admin
/// wrote, and whatever a future automatic rule stamps. This screen renders
/// every message the same way; the field exists so the two never need to
/// become different kinds of row.
class AppNotification {
  final int id;
  final String? title;
  final String body;
  final String source;
  final bool isRead;
  final DateTime? readAt;
  final DateTime createdAt;

  AppNotification({
    required this.id,
    required this.title,
    required this.body,
    required this.source,
    required this.isRead,
    required this.readAt,
    required this.createdAt,
  });

  factory AppNotification.fromJson(Map<String, dynamic> json) {
    return AppNotification(
      id: json['id'] as int,
      title: json['title'] as String?,
      body: json['body'] as String,
      source: json['source'] as String,
      isRead: json['is_read'] as bool,
      readAt: json['read_at'] == null ? null : DateTime.parse(json['read_at'] as String).toLocal(),
      createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
    );
  }
}

/// The inbox plus the one number the bell's indicator is drawn from -- the
/// backend counts unread messages, this app never does.
class NotificationInbox {
  final int unreadCount;
  final List<AppNotification> items;

  NotificationInbox({required this.unreadCount, required this.items});

  factory NotificationInbox.fromJson(Map<String, dynamic> json) {
    return NotificationInbox(
      unreadCount: json['unread_count'] as int,
      items: (json['items'] as List<dynamic>)
          .map((e) => AppNotification.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: backend/app/schemas/notification.py
from datetime import datetime

from pydantic import BaseModel, Field


class NotificationOut(BaseModel):
    """One message in a user's inbox.

    `source` is provenance -- "manual" for an admin message, and whatever
    a future automatic rule stamps. The client shows the same row either
    way; it is here so an admin can see why something was sent."""

    id: int
    title: str | None
    body: str
    source: str
    is_read: bool
    read_at: datetime | None
    created_at: datetime


class NotificationListOut(BaseModel):
    """The inbox plus the one number the bell's indicator is drawn from,
    so the client never has to count unread messages itself."""

    unread_count: int
    items: list[NotificationOut]


class UnreadCountOut(BaseModel):
    """What the bell's dot is drawn from, on its own so the app bar never
    has to fetch the whole inbox."""

    unread_count: int


class NotificationSendIn(BaseModel):
    """What an admin fills in to write to one user."""

    user_id: int
    title: str | None = Field(default=None, max_length=120)
    body: str = Field(min_length=1, max_length=2000)


class AdminNotificationOut(NotificationOut):
    """What the admin sees back: the same message, plus who received it,
    so a sent-messages list reads without a second lookup."""

    user_id: int
    user_login: str
